### PR TITLE
Implement "local" as a regular KUBERNETES_PROVIDER

### LINKED ISSUE #26937: Implement "local" as a regular KUBERNETES_PROVIDER
When discussing the issue #26453, we came to conclusion that we should get rid of `hack/local-cluster-up.sh` script in favor of `cluster/kube-up.sh` which will do the same if the value of KUBERNETES_PROVIDER env var is set to `local`.

That will increase consistency - after setting up the KUBERNETES_PROVIDER user will be able to do just `cluster/kube-up.sh` and `cluster/kubectl.sh` to operate on local k8s instance.

### PR DESCRIPTION
**Pull request by nhlfr**
Make a possibility to set up a local k8s cluster by:

```
export KUBERNETES_PROVIDER=local
./cluster/kube-up.sh
```

and deprecate the hack/local-up-cluster.sh script in favor of that
solution.

Since kube-up.sh default behavior is to run the cluster in background
and shut it down by kube-down.sh, this commit also brings the usage
of pidfiles for each component.

Fixes #26937

The example of usage:

```
$ sudo KUBERNETES_PROVIDER=local ./cluster/kube-up.sh  
... Starting cluster using provider: local
... calling verify-prereqs
TODO: verify-prereqs
... calling kube-up
curl: (7) Failed to connect to 127.0.0.1 port 8080: Connection refused
API SERVER port is free, proceeding...
Go version: go version go1.6.2 linux/amd64
+++ [0608 09:26:35] Building the toolchain targets:
    k8s.io/kubernetes/hack/cmd/teststale
+++ [0608 09:26:36] Building go targets for linux/amd64:
    cmd/kubectl
    cmd/hyperkube
+++ [0608 09:26:38] Placing binaries
Detected host and ready to start services.  Doing some housekeeping first...
Using GO_OUT /home/nhlfr/go/src/k8s.io/kubernetes/_output/local/bin/linux/amd64
Starting services now!
Starting etcd
etcd -addr 127.0.0.1:4001 -data-dir /tmp/tmp.s92jyRcYaQ --bind-addr 127.0.0.1:4001 >/dev/null 2>/dev/null
Waiting for etcd to come up.
+++ [0608 09:26:39] On try 2, etcd: : http://127.0.0.1:4001
{"action":"set","node":{"key":"/_test","value":"","modifiedIndex":4,"createdIndex":4}}
Waiting for apiserver to come up
+++ [0608 09:26:40] On try 2, apiserver: : {
  "kind": "PodList",
  "apiVersion": "v1",
  "metadata": {
    "selfLink": "/api/v1/pods",
    "resourceVersion": "10"
  },
  "items": []
}
Local Kubernetes cluster is running. Press Ctrl-C to shut it down.

Logs:
  /tmp/kube-apiserver.log
  /tmp/kube-controller-manager.log
  /tmp/kube-proxy.log
  /tmp/kube-scheduler.log
  /tmp/kubelet.log

To start using your cluster, open up another terminal/tab and run:

  cluster/kubectl.sh config set-cluster local --server=http://127.0.0.1:8080 --insecure-skip-tls-verify=true
  cluster/kubectl.sh config set-context local --cluster=local
  cluster/kubectl.sh config use-context local
  cluster/kubectl.sh
... calling validate-cluster
Waiting for 1 ready nodes. 0 ready nodes, 0 registered. Retrying.
Found 1 node(s).
NAME        STATUS    AGE
127.0.0.1   Ready     14s
Validate output:
NAME                 STATUS    MESSAGE              ERROR
scheduler            Healthy   ok                   
controller-manager   Healthy   ok                   
etcd-0               Healthy   {"health": "true"}   
Cluster validation succeeded
Done, listing cluster services:

Kubernetes master is running at http://localhost:8080

To further debug and diagnose cluster problems, use 'kubectl cluster-info dump'.

$ sudo KUBERNETES_PROVIDER=local ./cluster/kube-down.sh
Bringing down cluster using provider: local
TODO: verify-prereqs
Cleaning up...
```

The example of using the deprecated `local-up-cluster.sh`:

```
$ sudo ./hack/local-up-cluster.sh
WARNING : This script is deprecated. Now the "local" KUBERNETES_PROVIDER is implemented, so to set up local cluster, use the following commands:

  export KUBERNETES_PROVIDER=local
  cluster/kube-up.sh

And to shut down the cluster:

  cluster/kube-down.sh

... Starting cluster using provider: local
... calling verify-prereqs
TODO: verify-prereqs
... calling kube-up
curl: (7) Failed to connect to 127.0.0.1 port 8080: Connection refused
API SERVER port is free, proceeding...
Go version: go version go1.6.2 linux/amd64
+++ [0608 09:47:49] Building the toolchain targets:
    k8s.io/kubernetes/hack/cmd/teststale
+++ [0608 09:47:49] Building go targets for linux/amd64:
    cmd/kubectl
    cmd/hyperkube
+++ [0608 09:47:51] Placing binaries
Detected host and ready to start services.  Doing some housekeeping first...
Using GO_OUT /home/nhlfr/go/src/k8s.io/kubernetes/_output/local/bin/linux/amd64
Starting services now!
Starting etcd
etcd -addr 127.0.0.1:4001 -data-dir /tmp/tmp.XGTNbPzJit --bind-addr 127.0.0.1:4001 >/dev/null 2>/dev/null
Waiting for etcd to come up.
+++ [0608 09:47:52] On try 2, etcd: : http://127.0.0.1:4001
{"action":"set","node":{"key":"/_test","value":"","modifiedIndex":4,"createdIndex":4}}
Waiting for apiserver to come up
+++ [0608 09:47:53] On try 2, apiserver: : {
  "kind": "PodList",
  "apiVersion": "v1",
  "metadata": {
    "selfLink": "/api/v1/pods",
    "resourceVersion": "10"
  },
  "items": []
}
Local Kubernetes cluster is running. Press Ctrl-C to shut it down.

Logs:
  /tmp/kube-apiserver.log
  /tmp/kube-controller-manager.log
  /tmp/kube-proxy.log
  /tmp/kube-scheduler.log
  /tmp/kubelet.log

To start using your cluster, open up another terminal/tab and run:

  cluster/kubectl.sh config set-cluster local --server=http://127.0.0.1:8080 --insecure-skip-tls-verify=true
  cluster/kubectl.sh config set-context local --cluster=local
  cluster/kubectl.sh config use-context local
  cluster/kubectl.sh
```

### CODE REVIEW

File: hack/local-up-cluster.sh [hunk 1]
```diff
@@ -14,41 +14,23 @@
...
     echo "WARNING : This script MAY be run as root for docker socket / iptables functionality; if failures occur, retry as root." 2>&1
 fi
 
-# Stop right away if the build fails
-set -e
+cat <<EOF
+WARNING : This script is deprecated. Now the "local" KUBERNETES_PROVIDER is implemented, so to set up local cluster, use the following commands:
+
+  export KUBERNETES_PROVIDER=local
+  cluster/kube-up.sh
```
Comment (dims) [inline hack/local-up-cluster.sh]: Neat! Thanks @nhlfr

**Comment by spxtr** [comment]
I like it!

There's a decent amount of documentation pointing at `local-up-cluster.sh` that will need to be fixed as well.

**Comment by pmorie** [comment]
I wrote most of the containerization stuff in `local-up-cluster.sh` and have felt like the env var names I picked were janky ever since.  If we really are going to deprecate `local-up-cluster.sh`, we can decide on whether we want to keep any of the old names or not.

With that said,here's a of history: there used to be a local provider, and it was taken out and `local-up-cluster.sh` made the true way.  I'm totally cool with `local-up-cluster.sh` being deprecated, but if we do that we need to make a commitment to maintaining the local provider.  I don't even know which sig to tag in for this...

**Comment by idvoretskyi** [comment]
@pmorie yes, I agree with you. @nhlfr made amazing work, but we have to save the compatibility with the work, that has been already implemented.

cc @kubernetes/sig-config to review.

**Comment by nhlfr** [comment]
@pmorie What concrete env variables you mean? From what I see, the only variables that strongly differ from convention that the other providers follow, are `API_HOST` and `KUBELET_HOST` (which may be replaces by `MASTER_IP` and `MASTER`). Anything else?
Unfortunately, I also don't see any concrete SIG which we may assign for maintaining this. But what SIG(s) is(are) responsible for maintaining the other providers? Maybe the same group of people may be assigned to maintaining local provider as well?
@idvoretskyi My current implementation is 100% backwards compatible.

**Comment by smarterclayton** [comment]
People typically use local-up-cluster for local dev, rather than for simulating a cluster locally.  I'm hesitant to say that we want to remove local-up-cluster, but I'm fine to add a 'local' provider implementation.

**Comment by smarterclayton** [comment]
Looking at your change, this makes local *look* like the other providers, but it is still very different.  All other cluster ups use the standard build process and the standard deployment.  This is substantially different, and may confuse users who expect the other options to be fixed.  hack/local-up-cluster doesn't *have* to be consistent with other providers.

**Comment by derekwaynecarr** [comment]
If we are going to do all this work, can we launch things properly on host systems.

For example, on systemd can we just start transient units similar to what was just done for node e2e?

see: https://github.com/kubernetes/kubernetes/pull/27167

**Comment by derekwaynecarr** [comment]
I thought we were deprecating the cluster directory?  /cc @mikedanese

**Comment by kargakis** [comment]
Please don't break existing behavior of ./hack/local-cluster-up.

**Comment by nhlfr** [comment]
This patch seems to be too controversial. I don't really have any interest in working on that, it just was a brief idea coming from some other discussion. So I'm closing it, feel free to re-open if someone wants that.

**Pull request closed without merging**